packages/event-handler/src/rest/RouteHandlerRegistry.ts

```diff
@@ -1,19 +1,66 @@
 import type { GenericLogger } from '@aws-lambda-powertools/commons/types';
-import type { RouteRegistryOptions } from '../types/rest.js';
+import type {
+  DynamicRoute,
+  HttpMethod,
+  Path,
+  RouteHandlerOptions,
+  RouteRegistryOptions,
+} from '../types/rest.js';
+import { ParameterValidationError } from './errors.js';
 import type { Route } from './Route.js';
-import { validatePathPattern } from './utils.js';
+import {
+  compilePath,
+  processParams,
+  validateParams,
+  validatePathPattern,
+} from './utils.js';
 
 class RouteHandlerRegistry {
-  readonly #routes: Map<string, Route> = new Map();
+  readonly #staticRoutes: Map<string, Route> = new Map();
+  readonly #dynamicRoutesSet: Set<string> = new Set();
+  readonly #dynamicRoutes: DynamicRoute[] = [];
   readonly #routesByMethod: Map<string, Route[]> = new Map();
+  #shouldSort = true;
 
   readonly #logger: Pick<GenericLogger, 'debug' | 'warn' | 'error'>;
 
   constructor(options: RouteRegistryOptions) {
     this.#logger = options.logger;
   }
 
+  /**
+   * Compares two dynamic routes to determine their specificity order.
+   * Routes with fewer parameters and more path segments are considered more specific.
+   * @param a - First dynamic route to compare
+   * @param b - Second dynamic route to compare
+   * @returns Negative if a is more specific, positive if b is more specific, 0 if equal
+   */
+  #compareRouteSpecificity(a: DynamicRoute, b: DynamicRoute): number {
+    // Routes with fewer parameters are more specific
+    const aParams = a.paramNames.length;
+    const bParams = b.paramNames.length;
+
+    if (aParams !== bParams) {
+      return aParams - bParams;
+    }
+
+    // Routes with more path segments are more specific
+    const aSegments = a.path.split('/').length;
+    const bSegments = b.path.split('/').length;
+
+    return bSegments - aSegments;
+  }
+  /**
+   * Registers a route in the registry after validating its path pattern.
+   *
+   * The function decides whether to store the route in the static registry
+   * (for exact paths like `/users`) or dynamic registry (for parameterized
+   * paths like `/users/:id`) based on the compiled path analysis.
+   *
+   * @param route - The route to register
+   */
   public register(route: Route): void {
+    this.#shouldSort = true;
```

Comment:
This handles the edge case where for some reason someone registers a route after resolve has been called.

Reply:
Good to have it, but I agree that this is not a pattern we should encourage